lib/tasks/npm-task.js

```diff
@@ -131,10 +143,11 @@ class NpmTask extends Task {
 
     logger.info('yarn.lock found -> trying yarn');
     try {
-      await this.checkYarn();
+      const { yarnVersion } = await this.checkYarn();
 
       logger.info('yarn found -> using yarn');
       this.useYarn = true;
+      this.yarnVersion = yarnVersion;
```

Comment:
I find it vaguely odd to set `this.yarnVersion` in two different ways. Can we refactor this a bit (possibly making `this.checkYarn` write the value itself) to avoid having to set the value twice based on control flow?

Reply:
Ah yeah looking back now it's hard to understand why I wrote it this way. I think I wanted to mirror the logic of returning an object of shape `{ xxxxVersion: version }` from `checkNpmVersion()`, but it also was necessary to set the `useYarn` flag.

Given that the only times `checkYarn()` is called is when `useYarn` is already `true` or when `yarn.lock` is found, I think it's fine to set the `useYarn` flag to `true` from inside `checkYarn()` right before returning. What I mean to say is that there is never a time that yarn is successfully found where we do not also want to use yarn.

The only exception I can think of is that someone has a project with a `yarn.lock` but is actually using `npm` to manage their packages, but I suspect that since yarn already warns about finding a `package-lock.json` & mixing package managers is pretty heavily discouraged, that's ok.

```js
try {
  const yarnVersionResult = await this.checkYarn();

  logger.info('yarn found -> using yarn');

  // if we get here, useYarn & yarnVersion are already set
  // this.useYarn = true;
  // this.yarnVersion = yarnVersion;

  // keeping this as a separate line from this.checkYarn() to allow logging
  return yarnVersionResult

} catch (_err) {
  logger.info('yarn not found -> using npm');
  return this.checkNpmVersion();
}
```

As a result, the only place that `this.yarnVersion` is set is right before the actual call to `yarn install`:

```js
if (this.useYarn) {
  this.yarnVersion = result.yarnVersion;
  let args = this.toYarnArgs(this.command, options);
  promise = this.yarn(args);
} else {
  let args = this.toNpmArgs(this.command, options);
  promise = this.npm(args);
// more stuff...
```